### PR TITLE
Rename skip-defaults, and add ability to exclude actual defaults

### LINKED ISSUE #523: dict(skip_defaults=True) doesn't apply to nested models
```python
from typing import List
from pydantic import BaseModel, Schema

DIGITAL_INPUT = 'digital_input'


class Payload(BaseModel):
    """Base class for all Combox payloads"""


class PointData(Payload):
    name: str = Schema(..., title='Name', min_length=1, max_length=32)
    gios_type: str = Schema(DIGITAL_INPUT, title='GIOS Type')


class PanelData(Payload):
    name: str = Schema(..., title='Name', min_length=1, max_length=32)
    points: List[PointData] = Schema(..., title='Points', min_length=1)


pd = PanelData.validate({'name': 'Dummy Panel', 'points': [{'name': 'Dummy Point', 'gios_type': DIGITAL_INPUT}]})
assert pd.dict(skip_defaults=True)['points'][0]['gios_type'] == DIGITAL_INPUT
print(pd.dict(skip_defaults=True))
```

```
{'name': 'Dummy Panel', 'points': [{'name': 'Dummy Point', 'gios_type': 'digital_input'}]}
```

Notice how the `dict` contains 'gios_type' field.

### PR DESCRIPTION
**Pull request by dmontagu**
## Change Summary

Rename `skip_defaults` to `exclude_unset`, and add `exclude_defaults` argument.

Still need to add tests and docs, but I wanted to get some feedback on the approach first.

## Related issue number

fix #523

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -584,34 +602,45 @@ def _iter(
...
 
         value_exclude = ValueItems(self, exclude) if exclude else None
         value_include = ValueItems(self, include) if include else None
 
+        if exclude_defaults:
+            if allowed_keys is None:
+                allowed_keys = set(self.__fields__)
+            for k in allowed_keys:
+                if self.__dict__[k] == self.__fields__[k].default:
+                    allowed_keys.discard(k)
```
Comment (dmontagu) [inline pydantic/main.py]: @samuelcolvin does the above logic seem reasonable to you for how to perform this? I initially put it in the `for k, v in self.__dict__.items()` loop, but this approach results in fewer checks when `exclude_defaults` is `False` (which I think will be the most common scenario).

Also, this feels like another small nudge toward storing the field defaults in their own dict for fast lookup. Given https://github.com/samuelcolvin/pydantic/pull/898, do you think it could make sense to add that?

I'm fine either way.

Comment (samuelcolvin) [reply pydantic/main.py]: this looks like it's duplicated below - presumably you meant that to demonstrate both approaches.

This route looks preferable to me.

---

I agree with you on `self.__defaults__` especially if the values are somehow shared between `self.__fields__[k].default` and `self.__defaults__[k]`. 

It would be as simple as adding:

```py
'__defaults__': {n: f.default for n, f in fields.items()},
```
to `new_namespace` in `ModelMetaclass`

Comment (dmontagu) [reply pydantic/main.py]: I didn't actually mean it in both cases 🤦‍♂️ hard to parse the diffs with all the renames. Thanks for all this feedback.

Comment (dmontagu) [reply pydantic/main.py]: @samuelcolvin how should this be handled if the field is required? Should I just put the non-required defaults in `__defaults__`? That's what we'd want for #898 too.

File: docs/usage/exporting_models.md [hunk 2]
```diff
@@ -10,7 +10,7 @@ Arguments:
 * `include`: fields to include in the returned dictionary; see [below](#advanced-include-exclude)
 * `exclude`: fields to exclude from the returned dictionary; see [below](#advanced-include-exclude)
 * `by_alias`: whether field aliases should be used as keys in the returned dictionary; default `False`
-* `skip_defaults`: whether fields which were not explicitly set when creating the model should
+* `exclude_unset`: whether fields which were not explicitly set when creating the model should
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/usage/exporting_models.md]: Done -- does the added note look good to you?

File: pydantic/main.py [hunk 3]
```diff
@@ -345,8 +358,20 @@ def json(
 
         `encoder` is an optional function to supply as `default` to json.dumps(), other arguments as per `json.dumps()`.
         """
+        if skip_defaults is not None:
+            warnings.warn(
+                f'{type(self).__name__}.dict: "skip_defaults" is deprecated and replaced by "exclude_unset"',
```
Comment (samuelcolvin) [inline pydantic/main.py]:
```suggestion
                f'{type(self).__name__}.json: "skip_defaults" is deprecated and replaced by "exclude_unset"',
```
?

File: pydantic/main.py [hunk 4]
```diff
@@ -303,15 +303,16 @@ def dict(
         include: Union['SetIntStr', 'DictIntStrAny'] = None,
         exclude: Union['SetIntStr', 'DictIntStrAny'] = None,
         by_alias: bool = False,
-        skip_defaults: bool = False,
+        exclude_unset: bool = False,
+        exclude_defaults: bool = False,
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/main.py]: I changed the default value to None so I can tell if it got set -- that way, you get a warning if you were specifying it as True *or* False (since, if you were specifying as `False`, it would start erroring if/when we removed it.) I think a more robust approach might be to pick up `**kwargs`, but that feels like overkill (and might be computationally expensive too).

I would imagine there are some neat decorators that could be used to perform this, but I'm also fine with the approach implemented here, or with leaving everything defaulting to False and just not warning people if they are passing `skip_defaults=False` somewhere. I leave it up to you

File: pydantic/main.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/main.py]: yes, `if not f.required` makes sense.

I just tried this on #898 and pycharm complained that `__defaults__` is a special name, see [here](https://docs.python.org/3/reference/datamodel.html). 

Do you think it's okay to call it that? I'm not at all clear whether it might break something.

File: pydantic/main.py (hunk 3 above)
Comment (dmontagu) [inline pydantic/main.py]: Does this warning look good to you?

Comment (samuelcolvin) [reply pydantic/main.py]:
```suggestion
                f'{self.__class__.__name__}.dict(): "skip_defaults" is deprecated and replaced by "exclude_unset"',
```
I would do this, but otherwise looks fine.

File: pydantic/main.py (hunk 1 above)
Comment (dmontagu) [reply pydantic/main.py]: I implemented this where `__defaults__` only holds the values for non-required fields. I think it's basically optimally fast when checking to exclude defaults -- you'd have to check them all, and they are cached. This should also simplify the logic for #898.

File: docs/usage/exporting_models.md [hunk 5]
```diff
@@ -10,7 +10,10 @@ Arguments:
...
+  be excluded from the returned dictionary; default `False`
+  * Prior to **v1.0**, `exclude_unset` was known as `skip_defaults`; use of `skip_defaults` is now deprecated
```
Comment (samuelcolvin) [inline docs/usage/exporting_models.md]: This isn't working because you need 4 spaces for mkdocs (:gun:), but also I think better to just have one long sentence than a sub-point.

File: pydantic/main.py (hunk 1 above)
Comment (dmontagu) [reply pydantic/main.py]: Good catch on the naming -- let's call it `__field_defaults__`? I'll rename it to that, but let me know if you'd prefer something else.

File: docs/usage/exporting_models.md [hunk 6]
```diff
@@ -10,7 +10,10 @@ Arguments:
 * `include`: fields to include in the returned dictionary; see [below](#advanced-include-exclude)
 * `exclude`: fields to exclude from the returned dictionary; see [below](#advanced-include-exclude)
 * `by_alias`: whether field aliases should be used as keys in the returned dictionary; default `False`
-* `skip_defaults`: whether fields which were not explicitly set when creating the model should
+* `exclude_unset`: whether fields which were not explicitly set when creating the model should
+  be excluded from the returned dictionary; default `False`.
+  Prior to **v1.0**, `exclude_unset` was known as `skip_defaults`; use of `skip_defaults` is now deprecated
```
Comment (dmontagu) [inline docs/usage/exporting_models.md]: Does this look better?

File: pydantic/main.py [hunk 7]
```diff
@@ -345,8 +358,20 @@ def json(
 
         `encoder` is an optional function to supply as `default` to json.dumps(), other arguments as per `json.dumps()`.
         """
+        if skip_defaults is not None:
+            warnings.warn(
+                f'{self.__class__.__name__}.dict(): "skip_defaults" is deprecated and replaced by "exclude_unset"',
```
Comment (samuelcolvin) [inline pydantic/main.py]:
```suggestion
                f'{self.__class__.__name__}.json(): "skip_defaults" is deprecated and replaced by "exclude_unset"',
```

**Comment by samuelcolvin** [comment]
need tests for the warning raise by `skip_defaults` and that it does the same thing as `exclude_defaults`

**Comment by samuelcolvin** [comment]
also need a change file.

**Comment by samuelcolvin** [comment]
I think this is the only thing v1 is waiting for. If possible it would be great if this could get finished.

Otherwise hope you don't if I push to this PR to finish it tomorrow.

**Comment by samuelcolvin** [comment]
awesome, thank you.
